tests/e2e/ctl_v3_auth_cluster_test.go

```diff
@@ -0,0 +1,111 @@
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.etcd.io/etcd/tests/v3/framework/config"
+	"go.etcd.io/etcd/tests/v3/framework/e2e"
+)
+
+func TestAuthCluster(t *testing.T) {
+	e2e.BeforeTest(t)
+	cfg := &e2e.EtcdProcessClusterConfig{
+		ClusterSize:   1,
+		InitialToken:  "new",
+		SnapshotCount: 2,
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	epc, err := e2e.NewEtcdProcessCluster(ctx, t, cfg)
+	if err != nil {
+		t.Fatalf("could not start etcd process cluster (%v)", err)
+	}
+	defer func() {
+		if err := epc.Close(); err != nil {
+			t.Fatalf("could not close test cluster (%v)", err)
+		}
+	}()
+
+	epcClient := epc.Client()
+	createUsers(ctx, t, epcClient)
+
+	if err := epcClient.AuthEnable(ctx); err != nil {
+		t.Fatalf("could not enable Auth: (%v)", err)
+	}
+
+	testUserClientOpts := e2e.WithAuth("test", "testPassword")
+	rootUserClientOpts := e2e.WithAuth("root", "rootPassword")
+
+	// write more than SnapshotCount keys to single leader to make sure snapshot is created
+	for i := 0; i <= 10; i++ {
+		if err := epc.Client(testUserClientOpts).Put(ctx, fmt.Sprintf("/test/%d", i), "test", config.PutOptions{}); err != nil {
+			t.Fatalf("failed to Put (%v)", err)
+		}
+	}
+
+	// start second process
+	if err := epc.StartNewProc(ctx, t, rootUserClientOpts); err != nil {
+		t.Fatalf("could not start second etcd process (%v)", err)
+	}
+
+	// make sure writes to both endpoints are successful
+	endpoints := epc.EndpointsV3()
+	assert.Equal(t, len(endpoints), 2)
+	for _, endpoint := range epc.EndpointsV3() {
+		if err := epc.Client(testUserClientOpts, e2e.WithEndpoints([]string{endpoint})).Put(ctx, "/test/key", endpoint, config.PutOptions{}); err != nil {
+			t.Fatalf("failed to write to Put to %q (%v)", endpoint, err)
+		}
+	}
+
+	// verify all nodes have exact same revision and hash
+	assert.Eventually(t, func() bool {
+		hashKvs, err := epc.Client(rootUserClientOpts).HashKV(ctx, 0)
+		if err != nil {
+			t.Logf("failed to get HashKV: %v", err)
+			return false
+		}
+		if len(hashKvs) != 2 {
+			t.Logf("not exactly 2 hashkv responses returned: %d", len(hashKvs))
+			return false
```

Comment:
It should be an error or fatal error, and fail the test immediately.

Reply:
In case of a node returns an error (which sometimes happens on first request) e2e hashkv implementation returns single entry without an explicit error.
Unfortunately  I'm not ready to troubleshoot that further as part of this PR, we've already spent quite a bit on finding issues with e2e test implementations.

Reply:
Maybe return code is not checked around here
https://github.com/etcd-io/etcd/blob/e5790d204c6a3d19941cb11a1d7d9b8901fee81f/pkg/expect/expect.go#L57-L75

Reply:
You already considered the case above, line 71 & 72

Reply:
71...72 will be triggered if HashKV() returns an error, but it does not, just a truncated response.

Reply:
It seems that we might need to enhance [(*ExpectProcess) ExpectFunc](https://github.com/etcd-io/etcd/blob/c45f3388ac93f9645b5dea4bcfea7020622207d3/pkg/expect/expect.go#L120). When the process exited with an error code ( != 0), then we need to return an error. 

Please feel free to investigate & enhance it in a separate PR. Let's update this PR per https://github.com/etcd-io/etcd/pull/14574/files#r1007400980

I expect any methods of [EtcdctlV3](https://github.com/etcd-io/etcd/blob/c45f3388ac93f9645b5dea4bcfea7020622207d3/tests/framework/e2e/etcdctl.go#L34), including `HashKV`, should return an error when the command `etcdctl` exits with an error code.

Reply:
I'm not sure I'll have time to work on ExpectProcess fixes, I don't know how many tests will fail if we change the behavior and what else will need to be updated.
I'll be happy to rebase this PR and update the test to explicitly fail once e2e env is fixed, will leave is as is for now.

Reply:
Tracked it in https://github.com/etcd-io/etcd/issues/14638

Reply:
I might take a look a bit later this week, can't promise.